File: 2023/day2.py

with open("inputs/day2.txt") as file:
    puzzle = file.read().strip()

result = 0
for line in puzzle.strip().split('\n'):
    id_, rest = line.split(":")
    sets = rest.split(";")
    for s in sets:
        game_cubes = {"red": 0, "blue": 0, "green": 0}
        configs = s.split(",")
        for config in configs:
            n, cube = config.strip().split()
            game_cubes[cube] = int(n)
        if game_cubes["red"] > 12 or game_cubes["green"] > 13 or game_cubes["blue"] > 14:
            break
    else:
        result += int(id_.split()[1])

print(result)

result = 0
for line in puzzle.strip().split('\n'):
    id_, rest = line.split(":")
    sets = rest.split(";")
    game_cubes = {"red": [], "blue": [], "green": []}
    for s in sets:
        for config in s.split(","):
            n, cube = config.strip().split()
            game_cubes[cube].append(int(n))

    r, g, b = max(game_cubes["red"]), max(game_cubes["green"]), max(game_cubes["blue"])
    result += r*g*b


print(result)

File: 2023/day2.c
#include <stdio.h>
#include <string.h>

#define MAX_LINE 1000

int main() {
	FILE *fptr;
	char line[MAX_LINE];
	int result = 0;
	fptr = fopen("inputs/day2.txt", "r");
	if (fptr == NULL) {
		perror("Error in opening file day2.txt");
		return 1;
	}

	while ((fgets(line, MAX_LINE, fptr)) != NULL) {
		int number = 0;
		int blue = 0, red = 0, green = 0;
		int id = 0;
		int possible = 1;
		for (int i = 0; i < MAX_LINE && line[i] != '\0'; i++) {
			if (line[i] >= '0' && line[i] <= '9') {
				number = (line[i] - '0') + number * ((number!=0)?10:1);
				continue;
			} else if (line[i] == ':') {
				id = number;
				number = 0;
			} else if (line[i] == 'b') {
				blue = number;
				number = 0;
			} else if (line[i] == 'r') {
				red = number;
				number = 0;
			} else if (line[i] == 'g') {
				green = number;
				number = 0;
			} 
			if (red > 12 || green > 13 || blue > 14) {
				possible = 0;
				break;
			}
		}
		result += possible ? id : 0;
	}
	printf("Part one: %d\n", result);
    fseek(fptr, 0, SEEK_SET);

	// Part two
	result = 0;
	while ((fgets(line, MAX_LINE, fptr)) != NULL) {
		int number = 0;
		int blue = 0, red = 0, green = 0;
		int possible = 1;
		for (int i = 0; i < MAX_LINE && line[i] != '\0'; i++) {
			if (line[i] >= '0' && line[i] <= '9') {
				number = (line[i] - '0') + number * ((number!=0)?10:1);
				continue;
			} else if (line[i] == ':') {
				number = 0;
			} else if (line[i] == 'b') {
				blue = (number>blue)?number:blue;
				number = 0;
			} else if (line[i] == 'r') {
				red = (number>red)?number:red;
				number = 0;
			} else if (line[i] == 'g') {
				green = (number>green)?number:green;
				number = 0;
			}
		}
		result += red*blue*green;
	}
	printf("Part two: %d\n", result);
	fclose(fptr);
}